### PR TITLE
fix: Support pnpm patches in prune

### PR DESCRIPTION
**Pull request by chris-olszewski**
Addresses #1963 

This PR fixes two issues with the current implementation of prune for pnpm:
 - Pruning the importers section of `pnpm-lock.yaml`, previously the pruned lockfile would pickup any direct dependency changes for any project in the workspace
 - Copy dependency patches into the output directory while pruning

Fixing the first issue required updating the `Lockfile` interface take a list of workspace dependencies when constructing the subgraph. This is required as workspace dependencies don't appear in `packages` in the pnpm lockfile format. Taking a quick look at the npm lockfile format, this additional information will be helpful when implementing npm prune so it seems sensible to make this addition.

The current implementation of copying dependency patches copies all patches instead of those that are strictly necessary. This was done due to the fact that `pnpm` duplicates patch info in `package.json` and in order to only copy the necessary patches we would first need to learn how to prune `package.json`. This should be done in the future, but is outside of the scope of this fix.

### CODE REVIEW

File: cli/internal/lockfile/pnpm_lockfile.go [hunk 1]
```diff
@@ -180,20 +181,52 @@ func (p *PnpmLockfile) Subgraph(packages []string) (Lockfile, error) {
...
+func pruneImporters(importers *map[string]ProjectSnapshot, workspacePackages []turbopath.AnchoredSystemPath) (map[string]ProjectSnapshot, error) {
+	prunedImporters := map[string]ProjectSnapshot{}
+
+	// Copy over root level importer
+	if root, ok := (*importers)["."]; ok {
+		prunedImporters["."] = root
+	}
+
+	for _, workspacePath := range workspacePackages {
+		workspace := workspacePath.ToUnixPath().ToString()
```
Comment (chris-olszewski) [inline cli/internal/lockfile/pnpm_lockfile.go:217]: pnpm encodes workspace paths using `/` as a separator regardless of the separator the OS uses

**Review by gsoltis — APPROVED** [review]

**Review by nathanhammond — APPROVED** [review]
This is an approval contingent on the patch location within workspaces. If that comes up as "not root every time" then I'd like another look.

File: cli/internal/lockfile/pnpm_lockfile.go [hunk 2]
```diff
@@ -180,20 +181,52 @@ func (p *PnpmLockfile) Subgraph(packages []string) (Lockfile, error) {
...
+		// before we can implement this we need to be able to prune the pnpm section
+		// of package.json otherwise installation will fail
+		PatchedDependencies: p.PatchedDependencies,
 	}
 
 	return &lockfile, nil
 }
 
+// Prune imports to only those have all of their dependencies in the packages list
+func pruneImporters(importers *map[string]ProjectSnapshot, workspacePackages []turbopath.AnchoredSystemPath) (map[string]ProjectSnapshot, error) {
```
[the remark this answers is not in the document]
Comment (chris-olszewski) [reply cli/internal/lockfile/pnpm_lockfile.go:208]: Ah, good catch.

File: cli/internal/lockfile/pnpm_lockfile.go [hunk 3]
```diff
@@ -209,9 +242,23 @@ func (p *PnpmLockfile) Encode(w io.Writer) error {
...
 }
 
+// Patches return a list of patches used in the lockfile
+func (p *PnpmLockfile) Patches() []string {
+	if len(p.PatchedDependencies) == 0 {
+		return nil
+	}
+	patches := make([]string, len(p.PatchedDependencies))
+	i := 0
+	for _, patch := range p.PatchedDependencies {
```
[the remark this answers is not in the document]
Comment (chris-olszewski) [reply cli/internal/lockfile/pnpm_lockfile.go:252]: Since `PatchedDependencies` is a map wouldn't `i` become the key in that case?

File: cli/internal/prune/prune.go [hunk 4]
```diff
@@ -190,10 +190,19 @@ func (p *prune) prune(opts *opts) error {
...
 
-	lockfile, err := ctx.Lockfile.Subgraph(lockfileKeys)
+	lockfile, err := ctx.Lockfile.Subgraph(workspaces, lockfileKeys)
 	if err != nil {
 		return errors.Wrap(err, "Failed creating pruned lockfile")
 	}
+
+	if patches := lockfile.Patches(); patches != nil {
+		for _, patch := range patches {
+			if err := fs.CopyFile(&fs.LstatCachedFile{Path: p.base.RepoRoot.Join(patch)}, fullDir.Join(patch).ToStringDuringMigration()); err != nil {
```
[the remark this answers is not in the document]
Comment (chris-olszewski) [reply cli/internal/prune/prune.go:200]: As of pnpm 7.11.0 conflicting patches just clobber any existing patch

File: cli/internal/lockfile/pnpm_lockfile.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gsoltis) [reply cli/internal/lockfile/pnpm_lockfile.go:251]: Ah, whoops, I missed the type of `PatchedDependencies`. You're right.

File: cli/internal/prune/prune.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nathanhammond) [reply cli/internal/prune/prune.go]: Let me rephrase, we're loading the patches from which of these locations?
- `<repo-root>/`
- `<repo-root>/packages/ui-library/`

Since we can define patches at any place in the workspace, I don't know how that resolves into the root lockfile:
```
{
  "pnpm": {
    "patchedDependencies": {
      "express@4.18.1": "patches/express@4.18.1.patch"
    }
  }
}
```

Does it resolve fully-qualified to the workspace path + patch path?

[the remark this answers is not in the document]
Comment (chris-olszewski) [reply cli/internal/prune/prune.go]: We should be loading from the repo root. `pnpm patch-commit` places patches in the workspace root and the paths are encoded in the lockfile relative to the repo root. pnpm hoists every patch and applies it to the entire workspace

[the remark this answers is not in the document]
Comment (chris-olszewski) [reply cli/internal/prune/prune.go]: I'll also add that `pnpm` doesn't respect patches defined in sub-packages so if for example you had your patch listed in `packages/ui-library/package.json` the patch won't appear in the lockfile.

[the remark this answers is not in the document]
Comment (weyert) [reply cli/internal/prune/prune.go]: Code suggests its always in the same directory as the lock file?
https://github.com/pnpm/pnpm/blob/main/packages/plugin-commands-patching/src/patchCommit.ts#L35

[the remark this answers is not in the document]
Comment (nathanhammond) [reply cli/internal/prune/prune.go]: Thank y'all for explaining how it works!￼
